Consolidate dockerfile, update renv, and add tests

zkamvar opened the pull request:
This pull request achieves three things in one:

1. the dockerfile consolidation to fix #9 
2. updates the renv.lock file (see below)
3. adds a test to compare the latest vs the pull request version

You may ask yourself: why did Zhian shove three features into one PR? Well, I'll tell you.

## Dockerfile consolidation

The dockerfile consolidation worked pretty well with the only exception being: the image took >10 minutes to build instead of the usual 2 minutes. The reason for this is because the previous image was pinned to R 4.4.2 whereas the new image uses the latest version of R version 4 (which is 4.5). 

Because of this, several packages needed to be built from source. 

## updating the renv.lock file

My solution was to update the renv.lock file by using renv::restore() and renv::update() and renv::snapshot(). This updated all of the packages in the lockfile and the version of R itself. 

## Testing

The next challenge: how do I test that this behaves the same way as I previously expected? The answer was to write a test. Since the flu metrocast hub runs in about a minute, I set up a test to test that by [borrowing a test function from hubPredEvalsData](https://github.com/hubverse-org/hubPredEvalsData/blob/main/tests/testthat/helper-expect_df_equal_up_to_order.R) to overcome the fact that [the rows are not ordered](https://github.com/hubverse-org/hubPredEvalsData/issues/25).

The test does the following

1. fetch the flu metrocast
2. fetch the predevals-config from the dashboard repo
3. load the latest version of this image and generate data, output into a folder called latest
4. load the PR version of this image and generate data, output into a folder called new
5. use `scripts/test.R` from the container to compare the two outputs. It will error if they are not equal.

zkamvar committed:
first attempt at dockerfile consolidation

zkamvar committed:
add tests (and missing pkgs)

zkamvar committed:
update renv lock

This updates the renv lockfile wholly because the new R version was
causing problems

zkamvar committed:
test inside folder

zkamvar committed:
annotate testing

zkamvar committed:
don't descend into new dir in empty env

zkamvar commented on .github/workflows/build-container.yaml:
```diff
@@ -112,8 +149,43 @@ jobs:
+          # COMPARE OUTPUTS ---------------------------------------------------
+          echo "::group::[validation] comparing output data sets"
+          docker run --rm -i -v "$(pwd)":"/project" \
+            "ghcr.io/hubverse-org/hubpredevalsdata-docker:${TAG}" \
```
```suggestion
            "ghcr.io/hubverse-org/hubpredevalsdata-docker:latest" \
```

lshandross reviewed (approved):
Approved, but just to clarify: it is necessary for the tests that the evaluation data builds the same with both images twice; once during the build job and once during the test job?

zkamvar committed:
run second test from latest release

The idea behind this test is to prevent maliciously modifying the tests in the PR so they pass.

zkamvar commented:
> Approved, but just to clarify: it is necessary for the tests that the evaluation data builds the same with both images twice; once during the build job and once during the test job?

@lshandross 

Kind of. I just added [7e0e2ed](https://github.com/hubverse-org/hubPredEvalsData-docker/pull/11/commits/7e0e2edce6eeb4fe6023d1bbd9a489d289616c56), which makes the second test run from the latest release. 

The idea behind the two tests comes from [the dashboard site builder test strategy](https://docs.hubverse.io/en/latest/developer/dashboard-site.html#test). It is a way to balance the need for security and the need to update tests.

- The first test uses the tests embedded in the current PR
  - allows for updated expectations and more tests to be added 
- the second uses the tests from the current release
  - prevents malicious intent by confirming the test suite from the source container works
    as expected.
  - this is allowed to fail if the tests are updated for a valid reason (as in this case where there are no tests in the release).

lshandross commented:
Ok, thanks for the clarification

zkamvar commented:
One important note: the second tests ONLY run on pull requests and manual triggers of non-main branches. When publishing, the tests come from the suite that exists on the main branch (which ensures that the tests can be updated).
